ssl/statem/statem_srvr.c

```diff
@@ -1924,15 +1923,30 @@ static int tls_early_post_process_client_hello(SSL_CONNECTION *s)
         OSSL_TRACE_END(TLS_CIPHER);
     }
 
-    for (loop = 0; loop < clienthello->compressions_len; loop++) {
-        if (clienthello->compressions[loop] == 0)
-            break;
-    }
-
-    if (loop >= clienthello->compressions_len) {
-        /* no compress */
-        SSLfatal(s, SSL_AD_DECODE_ERROR, SSL_R_NO_COMPRESSION_SPECIFIED);
-        goto err;
+    /* Verify the compression methods */
+    if (SSL_CONNECTION_IS_TLS13(s)) {
+        /* >=TLSv1.3: RFC 8446 - only zero compression is allowed */
+        if (clienthello->compressions_len != 1) {
+            if (clienthello->compressions_len == 0)
+                SSLfatal(s, SSL_AD_DECODE_ERROR,
+                         SSL_R_NO_COMPRESSION_SPECIFIED);
+            else
+                SSLfatal(s, SSL_AD_ILLEGAL_PARAMETER,
+                         SSL_R_TLSV13_ONLY_NULL_COMPRESSION_ALLOWED);
+        }
+        if (clienthello->compressions[0] != 0) {
+            SSLfatal(s, SSL_AD_ILLEGAL_PARAMETER,
+                     SSL_R_TLSV13_ONLY_NULL_COMPRESSION_ALLOWED);
+            goto err;
+        }
+    } else {
+        /* <=TLSv1.2: Valid if zero compression is supported. */
+        if (memchr(clienthello->compressions, 0,
+                   clienthello->compressions_len) == NULL) {
+            /* No zero compression method found */
+            SSLfatal(s, SSL_AD_DECODE_ERROR, SSL_R_NO_COMPRESSION_SPECIFIED);
+            goto err;
+        }
```

Comment:
that's code duplication, isn't it?

wouldn't it be better to fist check if there's at least one value there, irrespective of protocol version, and then if we have negotiated TLS1.3 abort it there's more than one or the one is anything but null?

Reply:
Yes, this is indeed duplicated code, but it was intentional for the sake of readability. The compiler will likely optimize the duplicated code away.

The `if` clause is designed to clearly separate the handling for TLSv1.3+ from the older behaviour. Moving the zero-length check outside of this `if` clause would blur that separation and still require a check for null compression in the array.

Reply:
Here’s the code for comparison:

```c
    /* Verify the compression methods */
    if (clienthello->compressions_len == 0) {
        SSLfatal(s, SSL_AD_DECODE_ERROR,
                 SSL_R_NO_COMPRESSION_SPECIFIED);
        goto err;
    }
    if (SSL_CONNECTION_IS_TLS13(s)) {
        /* >=TLSv1.3: RFC 8446 - only zero compression is allowed */
        if (clienthello->compressions_len > 1) {
            SSLfatal(s, SSL_AD_ILLEGAL_PARAMETER,
                     SSL_R_TLSV13_ONLY_NULL_COMPRESSION_ALLOWED);
            goto err;
        }
        if (clienthello->compressions[0] != 0) {
            SSLfatal(s, SSL_AD_ILLEGAL_PARAMETER,
                     SSL_R_TLSV13_ONLY_NULL_COMPRESSION_ALLOWED);
            goto err;
        }
    } else {
        /* <=TLSv1.2: Valid if zero compression is supported. */
        if (memchr(clienthello->compressions, 0,
                   clienthello->compressions_len) == NULL) {
            /* No zero compression method found */
            SSLfatal(s, SSL_AD_DECODE_ERROR, SSL_R_NO_COMPRESSION_SPECIFIED);
            goto err;
        }
    }
```

In my opinion, this change doesn’t provide a significant improvement. However, if I misunderstood your original comment, please let me know.

Reply:
the TLS 1.3 ClientHello is by design backwards compatible with earlier versions, so having checks that are universal to all versions is the logical way to implement it

also, for TLS 1.2 we also should send illegal_parameter if the null compression method is missing:
```
      This vector MUST contain,
      and all implementations MUST support, CompressionMethod.null.
      Thus, a client and server will always be able to agree on a
      compression method.
```

Reply:
I’ve updated the code to correctly handle the pre-TLSv1.3 cases:

```c
    /* Verify the compression methods */
    if (clienthello->compressions_len == 0) {
        SSLfatal(s, SSL_AD_DECODE_ERROR, SSL_R_NO_COMPRESSION_SPECIFIED);
        goto err;
    }
    if (SSL_CONNECTION_IS_TLS13(s)) {
        /* >=TLSv1.3: RFC 8446 - only null compression is allowed */
        if (clienthello->compressions[0] != 0 ||
            clienthello->compressions_len > 1) {
            SSLfatal(s, SSL_AD_ILLEGAL_PARAMETER,
                     SSL_R_TLSV13_ONLY_NULL_COMPRESSION_ALLOWED);
            goto err;
        }
    } else {
        /* <=TLSv1.2: Valid if null compression is supported. */
        if (memchr(clienthello->compressions, 0,
                   clienthello->compressions_len) == NULL) {
            SSLfatal(s, SSL_AD_ILLEGAL_PARAMETER,
                     SSL_R_TLSV1_NULL_COMPRESSION_MISSING);
            goto err;
        }
    }
```

Now, only a zero-length compression method field triggers an `SSL_AD_DECODE_ERROR`, while all other cases trigger an `SSL_AD_ILLEGAL_PARAMETER` error. I also introduced a new error code, `SSL_R_TLSV1_NULL_COMPRESSION_MISSING`, to report the absence of null compression for TLSv1.2 and older protocols.